### ISSUE TITLE
FPS drop on running 2 video streams simultaneously

### ISSUE THREAD

**Issue by groot-1313** [AlexeyAB/darknet#1968]
Hi,
I have trained a tiny yolo v2 model.
I have not built darknet using opencv, so I cannot use ./darknet detector demo .. .
So, I am using the script darknet.py with a few modifications to run on a video file. I am timing only a single line to get the FPS -
detections = detect(netMain, metaMain, frame, thresh) 
I do this for 5000 frames and get the FPS as - total_frames/total_time_seconds .
When I run the python script from one terminal on a GTX 1080 TI, I get an FPS of 109 frames per second.
When I run the python script from 2 separate terminals simultaneously, they both print an FPS of 90 frames per second.

I would like to run 2 detection streams with no loss in FPS. How can I optimise further?

**Comment by groot-1313** [comment]
No, I am not getting 180 FPS in any scenario. 

Upon running a single instance of yolo inference on a video file, I get an FPS of 109. (1 GPU, 1 process)
Upon running two instances of yolo inference, I get an FPS of 90 on each of the processes. (1GPU, 2 processes)

**Comment by HagegeR** [comment]
90 FPS x 2 = 180

Le ven. 23 nov. 2018 à 11:08, groot-1313 <notifications@github.com> a
écrit :

**Comment by groot-1313** [comment]
So the real issue is that for each one of the processes the GPU isn't being completely used. Each process consumes about 350 MB of 4 GB GPU memory. If the GPU was being utilized completely for each process, and parallelly, then the FPS would have been 109 for each of the processes. 
Is this correct or am I missing something?

**Comment by HagegeR** [comment]
you should use GPU-Z or a similar program to see which part of the GPU is
your performance bottleneck here, RAM is likely not the problem here, using
more RAM won't get better performance in all cases.

Le ven. 23 nov. 2018 à 11:38, groot-1313 <notifications@github.com> a
écrit :

**Comment by sivaishere** [comment]
I have the same problem on the Tesla v100. However the fps gets drastically reduced. Running one instance of Tiny-yolo gives me 190fps. Running two instances reduces the fps to 95 per model. Running three instances of Tiny-yolo gives 60fps per model and so on. Shouldn't I b getting almost 190 fps on all three models. Is there an architecture difference in Tesla v100 that prohibits it from doing so? What seems to be the bottle neck here?

**Comment by sivaishere** [comment]
Running same experiment on RTX 2080ti, I was able to achieve 70fps running one instance. While running two models simultaneously i was able to achieve around 60 fps for both models. And while running five models i was able to achieve 40 fps on each one. I had to stop because my RAM was full.

How do I check if my gpu has attained the maximum number of operations per second? nvidia-smi gives GPU utilization. But as you said that is just gpu ram.

**Comment by HagegeR** [comment]
well in that case it means you'r bottleneck isn't the GPU cores, as I said earlier ([https://github.com/AlexeyAB/darknet/issues/1968#issuecomment-441210472](url)) you should profile your current test and see what part of the flow is your bottleneck, but even if you find it, solving it might mean needing to speed-up the code that @AlexeyAB has already optimized pretty well, or changing some hardware which is not always easy, or, if your lucky, changing some configuration.

**Comment by rod-hendricks** [comment]
I also notice in my tests that CPU% might be one of the bottleneck areas. I ran 9 python parallel instances of tiny yolo on RTX 2070 and get about 32FPS each with ~60% GPU utilization but 100% CPU utilization. I can't create more instances since i go into memory error already. Normally 1 python instance runs at 15% CPU usage while 1 darknet.exe instance runs at 30% CPU usage. Not exactly sure why the darknet.exe implementation consumes more CPU resources than python.

**Comment by AlexeyAB** [comment]
> Normally 1 python instance runs at 15% CPU usage while 1 darknet.exe instance runs at 30% CPU usage. 

May be darknet.exe gives 2x more FPS than python instance.

**Comment by rod-hendricks** [comment]
Yes that is true @AlexeyAB. Does it mean that the inference speed is very fast that the processing of images and rendering of output becomes the bottleneck which is why the CPU seems to be the bottleneck at least in Tiny Yolo?
